vortex-duckdb/src/convert/array/mod.rs

```diff
@@ -0,0 +1,155 @@
+use vortex_array::validity::Validity;
+use vortex_array::{Array, ArrayRef};
+use vortex_error::{VortexResult, vortex_err};
+
+use crate::convert::array::data_chunk_adaptor::{
+    DataChunkHandleSlice, NamedDataChunk, SizedFlatVector,
+};
+
+pub trait ToDuckDB {
+    fn to_duckdb(&self, chunk: &mut dyn WritableVector) -> VortexResult<()>;
+}
+
+pub fn to_duckdb_chunk(
+    struct_array: &StructArray,
+    chunk: &mut DataChunkHandle,
+) -> VortexResult<Vec<bool>> {
+    let mut nullable = vec![false; struct_array.len()];
+    for (idx, field) in struct_array.fields().iter().enumerate() {
+        field.to_duckdb(&mut DataChunkHandleSlice::new(chunk, idx))?;
+        nullable[idx] = field.dtype().is_nullable();
+    }
+    chunk.set_len(struct_array.len());
+    Ok(nullable)
+}
+
+impl ToDuckDB for ArrayRef {
+    fn to_duckdb(&self, chunk: &mut dyn WritableVector) -> VortexResult<()> {
+        let arrow = &self.clone().into_arrow_preferred()?;
+        write_arrow_array_to_vector(arrow, chunk)
+            .map_err(|e| vortex_err!("Failed to convert vrotex duckdb array: {}", e.to_string()))
+    }
+}
+
+pub trait FromDuckDB<V> {
+    fn from_duckdb(vector: V) -> VortexResult<ArrayRef>;
+}
+
+impl<'a> FromDuckDB<&'a NamedDataChunk<'a>> for ArrayRef {
+    fn from_duckdb(named_chunk: &'a NamedDataChunk<'a>) -> VortexResult<ArrayRef> {
+        let chunk = &named_chunk.chunk;
+        let names = &named_chunk.names;
+        let len = chunk.len();
+
+        let columns = (0..chunk.num_columns())
+            .map(|i| {
+                let vector = chunk.flat_vector(i);
+                let array = ArrayRef::from_duckdb(SizedFlatVector {
+                    vector,
+                    nullable: named_chunk.nullable.map(|null| null[i]).unwrap_or(true),
+                    len,
+                })?;
+
+                // Figure out the column names
+                Ok((
+                    names
+                        .as_ref()
+                        .map(|names| names[i].clone())
+                        .unwrap_or_else(|| i.to_string().into()),
+                    array,
+                ))
+            })
+            .collect::<VortexResult<Vec<_>>>()?;
+
+        let (names, arrays): (Vec<_>, Vec<_>) = columns.into_iter().unzip();
+
+        // all top level struct are non nullable is duckdb, only inner columns can be.
+        StructArray::try_new(names.into(), arrays, len, Validity::NonNullable)
+            .map(StructArray::into_array)
+    }
+}
+
+impl FromDuckDB<SizedFlatVector> for ArrayRef {
+    // TODO(joe): going via is slow, make it faster.
+    fn from_duckdb(mut sized_vector: SizedFlatVector) -> VortexResult<ArrayRef> {
+        let len = sized_vector.len;
+        let arrow_arr = flat_vector_to_arrow_array(&mut sized_vector.vector, len).map_err(|e| {
+            println!("Failed to convert duckdb duckdb array vortex: {}", e);
+
+            vortex_err!("Failed to convert duckdb array to vortex: {}", e)
```

Comment:
Why duplicate the error message in `println`?

Reply:
oops